R/fg_pip.R

```diff
@@ -30,61 +30,86 @@ fg_pip <- function(country,
+    ctry_years <- subset_lkup(country = country,
+                              year = year,
+                              welfare_type = welfare_type,
+                              reporting_level = reporting_level,
+                              lkup = interpolation_list[[unique_survey_files[svy_id]]]$ctry_years)
+
+    results_subset <- vector(mode = "list", length = nrow(ctry_years))
+
+    for (ctry_year_id in seq_along(ctry_years$interpolation_id)) {
+
+      # Extract records to be used for a single country-year estimation
+      tmp_metadata <- metadata[metadata$interpolation_id == ctry_years[["interpolation_id"]][ctry_year_id], ]
+
+      # Compute estimated statistics using the fill_gap method
+      if (debug) debugonce(wbpip:::prod_fg_compute_pip_stats)
+      tmp_stats <- wbpip:::prod_fg_compute_pip_stats(
+        request_year = ctry_years[["reporting_year"]][ctry_year_id],
+        data = svy_data,
+        predicted_request_mean = tmp_metadata[["predicted_mean_ppp"]],
+        svy_mean_lcu = tmp_metadata[["survey_mean_lcu"]],
+        survey_year = tmp_metadata[["survey_year"]],
+        default_ppp = tmp_metadata[["ppp"]],
+        ppp = ppp,
+        distribution_type = tmp_metadata[["distribution_type"]],
+        poverty_line = povline,
+        popshare = popshare
+      )
+
+      # Ensure that tmp_metadata has a single row
+      vars_to_collapse <- c(
+        "survey_id", "cache_id", "surveyid_year", "survey_year",
+        "survey_acronym", "survey_coverage", "survey_comparability",
+        "comparable_spell", "welfare_type", "distribution_type",
+        "gd_type", "mean", "predicted_mean_ppp", "survey_mean_lcu", "survey_mean_ppp",
+        "interpolation_id", "path", "cpi"
+      )
 
+      tmp_metadata[, vars_to_collapse] <- NA
 
-    # Add stats columns to data frame
-    for (j in seq_along(tmp_stats)) {
-      tmp_metadata[[names(tmp_stats)[j]]] <- tmp_stats[[j]]
+      tmp_metadata <- unique(tmp_metadata)
+
+
+      # Add stats columns to data frame
+      for (stat in seq_along(tmp_stats)) {
+        tmp_metadata[[names(tmp_stats)[stat]]] <- tmp_stats[[stat]]
+      }
+
+
+      # tmp_metadata <- collapse_rows(
+      #   df = tmp_metadata,
+      #   vars = vars_to_collapse,
+      #   na_var = "survey_mean_ppp"
+      # )
+      results_subset[[ctry_year_id]] <- tmp_metadata
     }
 
-    # Ensure that tmp_metadata has a single row
-    vars_to_collapse <- c(
-      "survey_id", "cache_id", "surveyid_year", "survey_year",
-      "survey_acronym", "survey_coverage", "survey_comparability",
-      "comparable_spell", "welfare_type", "distribution_type",
-      "gd_type", "predicted_mean_ppp", "survey_mean_lcu",
-      "interpolation_id", "path", "cpi"
-    )
-    tmp_metadata <- collapse_rows(
-      df = tmp_metadata,
-      vars = vars_to_collapse,
-      na_var = "survey_mean_ppp"
-    )
-    out[[i]] <- tmp_metadata
+    out[[svy_id]] <- results_subset
   }
 
+  out <- purrr::flatten(out)
```

Comment:
Maybe compare the speed of this with `unlist(out)` and specifically `unlist(out, use.names = FALSE)`.

Reply:
Thanks @Aeilert I don't think `unlist()` and `purrr::flatten()` are equivalent. `flatten` only removes one nesting level at a time when `unlist` removes all nesting levels (which is not what I want here)

Reply:
Ah, got it. I think you can use `recursive = FALSE` to achieve that.  Should be faster than `flatten`, especially with `use.names = FALSE`.

Reply:
I checked, and looks like `unlist(out, recurvise = FALSE)` gives identical results as `purrr::flatten(out)`. I implemented that change in an additional commit, so that we don't have to add purr as a dependency.